src/main/java/jcifs/pac/kerberos/KerberosEncData.java

```diff
@@ -0,0 +1,308 @@
+            break;
+        case KerberosConstants.RC4_ENC_TYPE:
+            decrypt = decryptRC4(data, key);
+            break;
+        default:
+            throw new GeneralSecurityException("Unsupported encryption type " + type);
+        }
+        return decrypt;
+    }
+
+
+    /**
+     * @param data
+     * @param key
+     * @return
+     * @throws GeneralSecurityException
+     * @throws NoSuchAlgorithmException
+     * @throws NoSuchPaddingException
+     * @throws InvalidKeyException
+     * @throws IllegalBlockSizeException
+     * @throws BadPaddingException
+     */
+    private static byte[] decryptRC4 ( byte[] data, Key key ) throws GeneralSecurityException, NoSuchAlgorithmException, NoSuchPaddingException,
+            InvalidKeyException, IllegalBlockSizeException, BadPaddingException {
+        Cipher cipher;
+        byte[] decrypt;
+        byte[] code = new byte[4];
+        Encdec.enc_uint32le(2, code, 0);
+        byte[] codeHmac = getHmac(code, key.getEncoded());
+
+        byte[] dataChecksum = new byte[KerberosConstants.CHECKSUM_SIZE];
+        System.arraycopy(data, 0, dataChecksum, 0, KerberosConstants.CHECKSUM_SIZE);
+
+        byte[] dataHmac = getHmac(dataChecksum, codeHmac);
+        SecretKeySpec dataKey = new SecretKeySpec(dataHmac, KerberosConstants.RC4_ALGORITHM);
+
+        cipher = Cipher.getInstance(KerberosConstants.RC4_ALGORITHM);
+        cipher.init(Cipher.DECRYPT_MODE, dataKey);
+
+        int plainDataLength = data.length - KerberosConstants.CHECKSUM_SIZE;
+        byte[] plainData = cipher.doFinal(data, KerberosConstants.CHECKSUM_SIZE, plainDataLength);
+
+        byte[] plainDataChecksum = getHmac(plainData, codeHmac);
+        if ( plainDataChecksum.length >= KerberosConstants.CHECKSUM_SIZE )
+            for ( int i = 0; i < KerberosConstants.CHECKSUM_SIZE; i++ )
+                if ( plainDataChecksum[ i ] != data[ i ] )
+                    throw new GeneralSecurityException("Checksum failed while decrypting.");
+
+        int decryptLength = plainData.length - KerberosConstants.CONFOUNDER_SIZE;
+        decrypt = new byte[decryptLength];
+        System.arraycopy(plainData, KerberosConstants.CONFOUNDER_SIZE, decrypt, 0, decryptLength);
+        return decrypt;
+    }
+
+
+    /**
+     * @param data
+     * @param key
+     * @param cipher
+     * @return
+     * @throws GeneralSecurityException
+     * @throws InvalidKeyException
+     * @throws InvalidAlgorithmParameterException
+     * @throws IllegalBlockSizeException
+     * @throws BadPaddingException
+     */
+    private static byte[] decryptDES ( byte[] data, Key key, Cipher cipher )
+            throws GeneralSecurityException, InvalidKeyException, InvalidAlgorithmParameterException, IllegalBlockSizeException, BadPaddingException {
+        byte[] decrypt;
+        try {
+            cipher = Cipher.getInstance("DES/CBC/NoPadding");
+        }
+        catch ( GeneralSecurityException e ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)